### PR TITLE
Rework / refactor of how the Give verb works. Give now has a re-bindable hotkey for it

### PR DESCRIPTION
**Pull request by SteelSlayer**
## What Does This PR Do
Reworks the Give verb (now named Give Item) to be point and click based. Adds a hotkey for the Give Item verb.

How it works now:
1. Put the item you want to give into your active hand.
2. Click the Give Item verb, or use the Give Item hotkey. Your mouse cursor and click behavior will change. Should the item be dropped, equipped, or should you switch hands, the click behavior will return to normal. Click the verb, or use your hotkey to return mouse behavior and icon to normal.
3. Clicking on a humanoid will offer your item directly to that person. This gives you an alert that lets you know you're in an "item offering state".
4. The person gets a HUD alert that shows the type of item they're offering. The alert briefly flashes to draw the receiver's attention to it.
5. The receiver can click the alert to accept the item. Or they may simply ignore it, and the give offer will expire after 10 seconds.

## Why It's Good For The Game
* Clicking a verb and scrolling through a list of people to select from can be clunky, especially with more than one person around (in Medbay for example).
* Having a re-bindable Give Item hotkey makes giving people items feel more fluid.
* The alert the receiver gets is less intrusive than the current alert popup in the middle of the screen.

## Images of changes
The mouse icon is basically a codersprite that I made, if anyone wants to supply a better/more colorful one, I will gladly use that over this one. This sprite has a 64x64 canvas size.

Mouse icon:
Sprite-0002

Giver alert that lets them know they're currently offering an item (made after the video below was recorded):
image

Demo:
https://user-images.githubusercontent.com/42044220/181870865-e811d7ea-f52b-4e2e-bbd3-e66ba261374e.mp4

## Testing
Used two separate accounts:
* Activated the Give Item verb and passed various items between the two mobs on each account. Tested the hotkey method, and the verb method. Worked in both cases.
* Clicked on things that shouldn't trigger the offering of the item (anything which is not a carbon mob). Verified that it doesn't try to give your item to a wall.
* Verified that dropping, equipping, switching hands, etc. to trigger the removal of the click override works.

## Changelog
:cl:
tweak: The Give verb has been reworked and renamed to Give Item. Activating "Give Item" will give you a different mouse icon, and change the behavior of your clicks. Whenever you click another humanoid player, you will offer the item to them, and your click behavior will return to normal.
add: Adds a "Give Item (Toggle)" hotkey players can use to activate "Give Item". The hotkey can be found under the Carbon section.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/give.dm [hunk 1]
```diff
@@ -1,52 +1,167 @@
...
+/datum/click_intercept/give/New(client/C)
+	..()
+	holder.mouse_pointer_icon = file("icons/mouse_icons/give_item.dmi")
+	to_chat(holder, "<span class='info'>You can now left click on someone to give them your held item.</span>")
+	RegisterSignal(holder.mob.get_active_hand(), list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), .proc/cancel_give)
+	RegisterSignal(holder.mob, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), .proc/cancel_give)
 
-	if(target.incapacitated() || usr.incapacitated() || target.client == null)
+
+/datum/click_intercept/give/Destroy(force = FALSE, silent = FALSE)
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:41]: `force` is unused

File: code/modules/mob/living/carbon/give.dm [hunk 2]
```diff
@@ -1,52 +1,167 @@
...
+	if(!receiver.client)
+		to_chat(user, "<span class='warning'>You offer [I] to [receiver], but they don't seem to respond...</span>")
+		return
+	if((receiver.r_hand && receiver.l_hand) || HAS_TRAIT(receiver, TRAIT_HANDS_BLOCKED))
+		to_chat(user, "<span class='warning'>[receiver] doesn't have any hands free to accept [I]!</span>")
+		return
+	// We use UID() here so that the receiver can have more then one give request at one time.
+	// Otherwise, throwing a new "give item" alert would override any current one also named "give item".
+	receiver.throw_alert("give item [I.UID()]", /obj/screen/alert/take_item, alert_args = list(user, receiver, I))
+	to_chat(user, "<span class='info'>You offer [I] to [receiver].</span>")
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:73]: you might need to add checks for consciousness.
And the total amount of hands.
(although maybe on the receiving side, unsure)

File: code/modules/mob/living/carbon/give.dm [hunk 3]
```diff
@@ -1,52 +1,167 @@
...
+	. = ..()
+	desc = "[giver] wants to hand you \a [I]. Click here to accept it!"
+	giver_UID = giver.UID()
+	receiver_UID = receiver.UID()
+	item_UID = I.UID()
+	add_overlay(icon(I.icon, I.icon_state, SOUTH))
+	add_overlay("alert_flash")
+	ADD_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
+	RegisterSignal(I, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), .proc/cancel_give)
+	RegisterSignal(giver, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), .proc/cancel_give)
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:106]: I feel bad about doing nontrivial gameplay changes in `/obj/screen/alert/Initialize()`. It should be just an rectangle on the screen; giving it traits to people feels weird.

However, i don't have a good idea how to fix this outside of a whole new `/datum/gifting_process` that would manage this, which feels terrifically overengineered (and probably worse of the two evils).
Sad.

File: code/modules/mob/living/carbon/give.dm [hunk 4]
```diff
@@ -1,52 +1,164 @@
...
+
+
+/obj/screen/alert/take_item/Destroy()
+	var/mob/living/giver = locateUID(giver_UID) // REMOVE_TRAIT needs a direct var reference.
+	REMOVE_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:111]: do you need to check that `giver` is not null here? (I don't know)

File: code/modules/mob/living/carbon/give.dm [hunk 5]
```diff
@@ -1,52 +1,164 @@
...
+	unregister_signals()
+	return ..()
+
+
+/obj/screen/alert/take_item/proc/unregister_signals()
+	UnregisterSignal(locateUID(giver_UID), list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS))
+	UnregisterSignal(locateUID(item_UID), list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:118]: same null check concern.

File: code/modules/mob/living/carbon/give.dm [hunk 6]
```diff
@@ -1,52 +1,164 @@
...
 
-	if(target.incapacitated() || usr.incapacitated() || target.client == null)
+
+/obj/screen/alert/take_item/proc/cancel_give()
+	SIGNAL_HANDLER
+	var/mob/living/giver = locateUID(giver_UID)
+	to_chat(giver, "<span class='warning'>You need to keep the item in your active hand if you want to hand it to someone!</span>")
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:124]: this one *definitely* needs a null check

File: code/modules/mob/living/carbon/give.dm [hunk 7]
```diff
@@ -1,52 +1,164 @@
...
+
+
+/obj/screen/alert/take_item/Click(location, control, params)
+	var/mob/living/receiver = locateUID(receiver_UID)
+	var/obj/item/I = locateUID(item_UID)
+	if(HAS_TRAIT(receiver, TRAIT_HANDS_BLOCKED))
+		to_chat(receiver, "<span class='warning'>You need to have your hands free to accept [I]!</span>")
+		return
+	var/mob/living/giver = locateUID(giver_UID)
+	if(receiver.r_hand && receiver.l_hand)
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:136]: null checks

File: code/modules/mob/living/carbon/give.dm [hunk 8]
```diff
@@ -1,52 +1,164 @@
...
+		to_chat(receiver, "<span class='warning'>You need to stay in reaching distance of [giver] to take [I]!</span>")
+		return
+	if((I.flags & NODROP))
+		to_chat(giver, "<span class='warning'>[I] stays stuck to your hand when [receiver] tries to take it!</span>")
+		to_chat(receiver, "<span class='warning'>[I] stays stuck to [giver]'s hand when you try to take it!</span>")
+		return
+	// Give is successful at this point, unregister signals so the dropped signal doesn't fire and call `cancel_give()`.
+	unregister_signals()
+	giver.unEquip(I)
+	receiver.put_in_hands(I)
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:149]: does this work sanely if the receiver is lacking hand(s)?

File: code/modules/mob/living/carbon/give.dm [hunk 9]
```diff
@@ -1,52 +1,164 @@
...
-				I.on_give(usr, target)
-			if("No")
-				target.visible_message("<span class='warning'> [usr.name] tried to hand [I] to [target.name] but [target.name] didn't want it.</span>")
-	else
-		to_chat(usr, "<span class='warning'> [target.name]'s hands are full.</span>")
+/obj/screen/alert/take_item/do_timeout(mob/M, category)
+	var/mob/living/giver = locateUID(giver_UID)
+	var/mob/living/receiver = locateUID(receiver_UID)
+	// Make sure we're still nearby. We don't want to show a message if the giver not near us.
+	if(get_dist(giver, receiver) <= 3)
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:160]: null checks.

File: code/datums/keybindings/carbon.dm [hunk 10]
```diff
@@ -22,6 +22,15 @@
 	var/mob/living/carbon/M = C.mob
 	M.throw_mode_off()
 
+/datum/keybinding/carbon/give_item
+	name = "Give Item (Toggle)"
```
Comment (moxian) [inline code/datums/keybindings/carbon.dm:26]: the semantics of the (toggle) is weird, because contrary to what i expected originally, if you:
- click the toggle (turning it on)
- click the person (giving them the prompt)
- click the toggle again 

then you won't STOP giving them the thing - you won't "toggle giving off".

Swapping hands turning gifting off makes sense, kinda, but it's a bit of a confusing UI.
Would it be possible to have the gift UI button have three (rather than two) states of outline - off / target selection /  waiting for the target to accept? And clicking on it cycling the mode as
- off -> target selection
- target selection -> off
- waiting -> off

(but i also expect that this is a bit too much works, so eh)

File: code/modules/mob/living/carbon/give.dm [hunk 11]
```diff
@@ -1,52 +1,164 @@
-/mob/living/carbon/verb/give(mob/living/carbon/target in oview(1))
-	set category = "IC"
-	set name = "Give"
+/**
+ * Toggles the [/datum/click_intercept/give] on or off for the src mob.
+ */
+/mob/living/carbon/proc/toggle_give()
+	if(HAS_TRAIT(src, TRAIT_OFFERING_ITEM))
+		to_chat(src, "<span class='warning'>You're already offering an item to someone!</span>")
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:6]: why can't you offer same item to multiple people tho? Would be both fun and practical, wouldn't it?
(although the code would be messy, i guess)

File: code/modules/mob/living/carbon/give.dm [hunk 12]
```diff
@@ -1,52 +1,164 @@
...
+	var/mob/living/carbon/receiver = object
+	var/obj/item/I = user.get_active_hand()
+	if(!user.Adjacent(object))
+		to_chat(user, "<span class='warning'>You need to be closer to [receiver] to offer them [I]</span>")
+		return
+	if(!receiver.client)
+		to_chat(user, "<span class='warning'>You offer [I] to [receiver], but they don't seem to respond...</span>")
 		return
+	if((receiver.r_hand && receiver.l_hand) || HAS_TRAIT(receiver, TRAIT_HANDS_BLOCKED))
+		to_chat(user, "<span class='warning'>[receiver] doesn't have any hands free to accept [I]!</span>")
```
Comment (moxian) [inline code/modules/mob/living/carbon/give.dm:68]: Of all the checks, i think this one is the most surmountable, and should not prevent you from trying to offer.
They will get the popup, click it, realize that their hands are full, fiddle with the inventory, and then click the popup again, and everyone would be happy.
But with this check here it becomes rather awkward - you need to verbally ask them to free their hands before you can offer them an item, which ruins the flow IMO.

File: code/modules/mob/living/carbon/give.dm (hunk 1 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:41]: Has to be there, since the parent has it.

File: code/modules/mob/living/carbon/give.dm (hunk 2 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:73]: Yeah, I think I need to add consciousness checks in multiple places actually.

File: code/modules/mob/living/carbon/give.dm (hunk 3 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:106]: Yep, I thought about using a datum briefly, but it just makes things more annoying.

File: code/modules/mob/living/carbon/give.dm (hunk 8 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:149]: Yes. There are hand checks further in the call stack for this proc.

File: code/modules/mob/living/carbon/give.dm (hunk 4 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:111]: I think I'm going to negate the need for null checks by registering `COMSIG_PARENT_QDELETING` to the giver and receiver. That way if either starts to be deleted, the alert is deleted first, and we know they exist.

File: code/modules/mob/living/carbon/give.dm (hunk 11 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:6]: It could be done, but yeah, I think the code would be a bit messy. I don't imagine too many situations where you would want to offer your item to multiple people at once. Usually you want to hand a singular person an item.

File: code/datums/keybindings/carbon.dm (hunk 10 above)
Comment (SteelSlayer) [reply code/datums/keybindings/carbon.dm:26]: I'm hoping players realize that after offering an item, having their normal mouse return along with the chat feedback means they're no longer in a giving state. Adding the third HUD icon state seems a bit odd.

File: code/modules/mob/living/carbon/give.dm (hunk 12 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:68]: Fair, I'll remove it.

File: code/modules/mob/living/carbon/give.dm (hunk 1 above)
Comment (moxian) [reply code/modules/mob/living/carbon/give.dm:41]: Oh, my bad; i checked the immediate `/datum/click_intercept/Destroy`, didn't found the arg, but didn't check the base `/datum/Destroy`

File: code/datums/keybindings/carbon.dm (hunk 10 above)
Comment (moxian) [reply code/datums/keybindings/carbon.dm:26]: > players realize that after offering an item, having their normal mouse return along with the chat feedback means they're no longer in a giving state.

They will realize, and that's the problem. Because they actually *are* still in a giving state until either the offer is accepted or the user swaps hands. But there's no visual distinction between this "giving state" and regular "just walking around doing ss13 things" one.
Also note how we you're using term "giving state" to mean "has click_intercept", whereas I use it to mean [effectively] `HAS_TRAIT(user, TRAIT_OFFERING_ITEM)`

Okay, constructively, UI-wise: How about the give button has only two states (on and off), but it becomes "off" not when the click_intercept is destroyed, but when the user loses TRAIT_OFFERING_ITEM / when `
/obj/screen/alert/take_item` is destroyed? (Or on manual click, which would do the same thing of canceling offer)

Comment (SteelSlayer) [reply code/datums/keybindings/carbon.dm:26]: I see what you're saying now. I think having the give icon also double as notification as to whether or not you're in a giving state might be confusing, I dunno. I'm thinking maybe I could make an alert for the giver which shows two people exchanging something as a representation that they're in a giving state.

**Comment by procdrone** [comment]
it always felt awkward to give items through right click menu, nice one!

File: code/modules/mob/living/carbon/give.dm [hunk 13]
```diff
@@ -1,52 +1,186 @@
...
+	to_chat(giver, "<span class='warning'>You need to keep the item in your active hand if you want to hand it to someone!</span>")
+	to_chat(locateUID(receiver_UID), "<span class='warning'>[giver] seems to have given up on giving you [locateUID(item_UID)].</span>")
+	qdel(src)
+
+
+/obj/screen/alert/take_item/Click(location, control, params)
+	var/mob/living/receiver = locateUID(receiver_UID)
+	if(receiver.stat != CONSCIOUS)
+		return
+	var/obj/item/I = locateUID(item_UID)
```
Comment (lewcc) [inline code/modules/mob/living/carbon/give.dm:156]: Again, will this still work sanely if the item has been deleted?

**Review by lewcc — COMMENTED** [review]
Minor things, overall looks really good

File: code/modules/mob/living/carbon/give.dm [hunk 14]
```diff
@@ -1,52 +1,186 @@
...
+
+
+/datum/click_intercept/give/InterceptClickOn(mob/user, params, atom/object)
+	if(user == object || !iscarbon(object))
+		return
+	var/mob/living/carbon/receiver = object
+	if(receiver.stat != CONSCIOUS)
+		to_chat(user, "<span class='warning'>[receiver] can't accept any items because they're [receiver == UNCONSCIOUS ? "unconscious" : "dead"]!</span>")
+		return
+	var/obj/item/I = user.get_active_hand()
```
[the remark this answers is not in the document]
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:75]: Yep, I'll just add `COMSIG_PARENT_QDELETING` to the list of things registered to the item.

File: code/modules/mob/living/carbon/give.dm (hunk 13 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:156]: This one I did have monitoring the item for `COMSIG_PARENT_QDELETING`. So as soon as the item is qdel'd this alert is qdel'd as well.

File: code/modules/mob/living/carbon/give.dm [hunk 15]
```diff
@@ -1,52 +1,186 @@
...
+	if(!user.Adjacent(receiver))
+		to_chat(user, "<span class='warning'>You need to be closer to [receiver] to offer them [I].</span>")
 		return
+	if(!receiver.client)
```
[the remark this answers is not in the document]
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:79]: I think it's too extra, but it could be a good idea yeah.

File: code/datums/datumvars.dm [hunk 16]
```diff
@@ -1176,7 +1176,7 @@
 		possibleverbs += typesof(/mob/proc,/mob/verb,/mob/living/proc,/mob/living/verb)
 		switch(H.type)
 			if(/mob/living/carbon/human)
-				possibleverbs += typesof(/mob/living/carbon/proc,/mob/living/carbon/verb,/mob/living/carbon/human/verb,/mob/living/carbon/human/proc)
+				possibleverbs += typesof(/mob/living/carbon/proc,/mob/living/carbon/human/verb,/mob/living/carbon/human/proc)
```
Comment (farie82) [inline code/datums/datumvars.dm:1179]: Why did you change this?

File: code/modules/mob/living/carbon/give.dm [hunk 17]
```diff
@@ -1,52 +1,184 @@
-/mob/living/carbon/verb/give(mob/living/carbon/target in oview(1))
-	set category = "IC"
-	set name = "Give"
+/**
+ * Toggles the [/datum/click_intercept/give] on or off for the src mob.
+ */
+/mob/living/carbon/proc/toggle_give()
+	if(has_status_effect(STATUS_EFFECT_OFFERING_ITEM))
+		to_chat(src, "<span class='warning'>You're already offering an item to someone!</span>")
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:6]: Design thing:
Maybe this should cancel the active offering instead?

File: code/modules/mob/living/carbon/give.dm [hunk 18]
```diff
@@ -1,52 +1,184 @@
...
+	UnregisterSignal(holder.mob.get_active_hand(), list(COMSIG_PARENT_QDELETING, COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
+	UnregisterSignal(holder.mob, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS))
+	return ..()
+
+
+/datum/click_intercept/give/InterceptClickOn(mob/user, params, atom/object)
+	if(user == object || !iscarbon(object))
+		return
+	var/mob/living/carbon/receiver = object
+	if(receiver.stat != CONSCIOUS)
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:72]: Probably won't happen... but this can be used to meta check against `TRAIT_FAKEDEATH`

File: code/modules/mob/living/carbon/give.dm [hunk 19]
```diff
@@ -1,52 +1,184 @@
...
+	RegisterSignal(I, COMSIG_PARENT_QDELETING, /datum/proc/signal_qdel)
+	RegisterSignal(giver, COMSIG_PARENT_QDELETING, /datum/proc/signal_qdel)
+	RegisterSignal(receiver, COMSIG_PARENT_QDELETING, /datum/proc/signal_qdel)
+
+
+/obj/screen/alert/take_item/Destroy()
+	var/mob/living/giver = locateUID(giver_UID)
+	giver.remove_status_effect(STATUS_EFFECT_OFFERING_ITEM)
+	if(!item_taken)
+		unregister_signals()
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:132]: Same here as earlier. I believe this is already handled by the datum implementation of `Destroy`

File: code/modules/mob/living/carbon/give.dm [hunk 20]
```diff
@@ -1,52 +1,184 @@
...
+		return
+	var/obj/item/I = locateUID(item_UID)
+	if(HAS_TRAIT(receiver, TRAIT_HANDS_BLOCKED) || receiver.r_hand && receiver.l_hand)
+		to_chat(receiver, "<span class='warning'>You need to have your hands free to accept [I]!</span>")
 		return
+	var/mob/living/giver = locateUID(giver_UID)
+	if(!giver.Adjacent(receiver))
+		to_chat(receiver, "<span class='warning'>You need to stay in reaching distance of [giver] to take [I]!</span>")
+		return
+	if((I.flags & NODROP))
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:162]:
```suggestion
	if(I.flags & NODROP)
```

File: code/modules/mob/living/carbon/give.dm [hunk 21]
```diff
@@ -1,52 +1,184 @@
...
+		to_chat(giver, "<span class='warning'>[I] stays stuck to your hand when [receiver] tries to take it!</span>")
+		to_chat(receiver, "<span class='warning'>[I] stays stuck to [giver]'s hand when you try to take it!</span>")
+		return
+	unregister_signals() // Give is successful at this point, unregister signals so the dropped signal doesn't fire and call `cancel_give()`.
+	item_taken = TRUE // This way we don't call `unregister_signals` again in Destroy.
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:167]: If the Destroy thing is true, then this shouldn't be required anymore

File: code/modules/mob/living/carbon/give.dm [hunk 22]
```diff
@@ -1,52 +1,184 @@
...
-				I.on_give(usr, target)
-			if("No")
-				target.visible_message("<span class='warning'> [usr.name] tried to hand [I] to [target.name] but [target.name] didn't want it.</span>")
-	else
-		to_chat(usr, "<span class='warning'> [target.name]'s hands are full.</span>")
+/obj/screen/alert/take_item/do_timeout(mob/M, category)
+	var/mob/living/giver = locateUID(giver_UID)
+	var/mob/living/receiver = locateUID(receiver_UID)
+	// Make sure we're still nearby. We don't want to show a message if the giver not near us.
+	if(get_dist(giver, receiver) <= 3)
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:180]: Design:
Maybe use a view check as well? Maybe they hid in a locker or such.

File: code/modules/mob/living/carbon/give.dm [hunk 23]
```diff
@@ -1,52 +1,184 @@
...
+	item_UID = I.UID()
+	giver.apply_status_effect(STATUS_EFFECT_OFFERING_ITEM)
+	add_overlay(icon(I.icon, I.icon_state, SOUTH))
+	add_overlay("alert_flash")
+	RegisterSignal(I, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), .proc/cancel_give)
+	RegisterSignal(giver, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), .proc/cancel_give)
+	// If any one of these atoms should be deleted, we need to cancel everything. Also saves having to do null checks before interacting with these atoms.
+	RegisterSignal(I, COMSIG_PARENT_QDELETING, /datum/proc/signal_qdel)
+	RegisterSignal(giver, COMSIG_PARENT_QDELETING, /datum/proc/signal_qdel)
+	RegisterSignal(receiver, COMSIG_PARENT_QDELETING, /datum/proc/signal_qdel)
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:125]: These shouldn't be required right? Since you clear the alerts when you qdel a mob now.
Will require some testing though

File: code/modules/mob/living/carbon/give.dm [hunk 24]
```diff
@@ -1,52 +1,184 @@
...
+	RegisterSignal(holder.mob, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), /datum/proc/signal_qdel)
+
+
+/datum/click_intercept/give/Destroy(force = FALSE, ...)
+	holder.mouse_pointer_icon = initial(holder.mouse_pointer_icon)
+	holder.mob.give_icon.icon_state = "act_give_off"
+	if(!item_offered)
+		to_chat(holder.mob, "<span class='info'>You're no longer trying to give someone your held item.</span>")
+	UnregisterSignal(holder.mob.get_active_hand(), list(COMSIG_PARENT_QDELETING, COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
+	UnregisterSignal(holder.mob, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS))
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:64]: These should already be handled right by the `Destroy` implementation of `/datum`?

**Review by farie82 — CHANGES_REQUESTED** [review]
Good stuff

File: code/datums/datumvars.dm (hunk 16 above)
Comment (SteelSlayer) [reply code/datums/datumvars.dm:1179]: So it turns out that was only one carbon verb, that being `give`. Removing it as a verb made this throw a compilation error.

File: code/modules/mob/living/carbon/give.dm (hunk 19 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:132]: Yeah, I forget that this stuff gets handled by datum Destroy. I'll clean all the unregisters up.

File: code/modules/mob/living/carbon/give.dm (hunk 23 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:125]: For the receiver, yeah I think I could remove it. Though the giver one is needed because the alert isn't attached to them.

File: code/modules/mob/living/carbon/give.dm (hunk 18 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:72]: Right, I'll have to account for that.

Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:72]: Alright so apparently when `TRAIT_FAKEDEATH` is added, they do eventually have their stat set to `UNCONSCIOUS` through a series of signal calls.

File: code/modules/mob/living/carbon/give.dm (hunk 17 above)
Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:6]: I'm not sure how I feel about this. However, what I will do make it so if the giver clicks the offering item alert, it cancels the offer.

File: code/modules/mob/living/carbon/give.dm (hunk 18 above)
Comment (lewcc) [reply code/modules/mob/living/carbon/give.dm:72]: Won't this still say that they're unconscious then, through the stat check in the message?

Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:72]: Yes.

**Comment by SteelSlayer** [comment]
I've just reverted all changes related to adding the new Give hud button. I realized that adding this new hud button would require multiple sprites for each hud theme (the red one, green one, white one, etc.). I'm not prepared to do that much work right now.

File: code/modules/mob/living/carbon/give.dm [hunk 25]
```diff
@@ -1,52 +1,189 @@
...
+ *
+ * While a mob has this intercept, left clicking on a carbon mob will attempt to offer their currently held item to that mob.
+ */
+/datum/click_intercept/give
+	/// If the intercept user has succesfully offered the item to another player.
+	var/item_offered = FALSE
+
+/datum/click_intercept/give/New(client/C)
+	..()
+	holder.mouse_pointer_icon = file("icons/mouse_icons/give_item.dmi")
```
Comment (farie82) [inline code/modules/mob/living/carbon/give.dm:73]:
```suggestion
	holder.mouse_pointer_icon = 'icons/mouse_icons/give_item.dmi'
```
Would this also work?

Comment (SteelSlayer) [reply code/modules/mob/living/carbon/give.dm:73]: It does indeed.

**Review by farie82 — APPROVED** [review]
